### ISSUE TITLE
[ARM] vm extension commands are extemely confusing

### ISSUE THREAD

**Issue by ahmetalpbalkan** [Azure/azure-xplat-cli#2397]
```
$ azure vm extension help
help:    Commands to manage VM resource extensions
help:      vm extension set [options] <resource-group> <vm-name> <name> <publisher-name> <version>
help:      vm extension get [options] <resource-group> <vm-name>
help:      vm extension list-image-publishers [options] <location>
help:      vm extension list-image-types [options] <location> <publisher>
help:      vm extension list-image-versions [options] <location> <publisher> <typeName>
help:      vm extension get-image [options] <location> <publisher> <typeName> <version>
```

May I ask how are `list-image-publishers` , `list-image-types`, `list-image-versions`, `get-image` are even related to VM Extensions? As a person publishing extensions, I have no idea what these commands are about.

When I run `list-image-publishers` it appears to be giving me VM Image publishers. Are we sure these commands should not be under `azure vm image`? This looks very odd to me.

Also there is no **`azure vm extension list`**? Where did  that go? It's very important actually.

@sauryadas

**Comment by huangpf** [comment]
The extension image cmdlets will be changed to ```vm extension-image list/show/list-publishers/list-versions```, which will be similar to the ```vm image``` pattern.

**Comment by ahmetalpbalkan** [comment]
@huangpf I don't think there's a thing called "extension images", it's extensions. This image thing is confusing. Do you mind telling where is that word coming from?

**Comment by huangpf** [comment]
https://github.com/Azure/azure-sdk-for-net/blob/master/src/ResourceManagement/Compute/ComputeManagement/Generated/VirtualMachineExtensionImageOperations.cs#L37

https://msdn.microsoft.com/en-us/library/mt150024.aspx

Note that the commands to adding/getting extensions to VMs remain the same, i.e. ```vm extension set/get ...```. The change is only about querying the source images of the extension published by different publishers that can be instantiated and then added to VMs, which is aligned with the current REST API, client library & PS cmdlet naming convention.

**Comment by squillace** [comment]
@huangpf The problem here is the word "images". As a user, I don't care much what the REST operation is; what I want to know is what extensions are available to me to use. That they extensions are "images" is not simply "merely" confusing, it's directly confusing. 

Is there a case in which there is a vm extension that is NOT an image? If so, then that might make sense. If there is no case in which that's possible, we need to **eliminate completely the word "image" from the end user API.

I don't know why we name something for what PS called it, when the description for the item is that thing, without the word "image". Snipped from the docs you cited:

```
Get-AzureVMExtensionImage
Gets all versions for an Azure extension.
```

????

**Comment by huangpf** [comment]
Hi @squillace,

Thanks for the feedback. @sauryadas are welcome to add any comments.

But in my opinion, extension images (a term introduced by the extension service API team, not PS/CLI team) refer to what *types* of extensions that can be added to VMs (i.e. BGInfo, CustomScript, etc.) alongside with configuration information, like version, config input, etc. They are similar to gallery items.

On the other hand, extensions (or I would say VM extensions) refer to the ones that users chose to instantiate and then are added to the VM with their customized configurations. They are more like applications running on the VMs.

So they are indeed different, but I understand that the word "images" are the center of debate here.

Options:
1) hide the key word that is introduced by the service team, and then keep using 'extensions' everywhere;
2) reword it to 'available extensions' from 'extension images', like what ASM commands did;
3) keep the concept aligned with the service API design, and use 'extension' vs 'extension images'

The 1 & 2 options will also introduce confusion. For example, does Get-AzureVMExtension mean getting from the VM or the published gallery? Option 3 is aligned with the ultimate goal to unify the design and usage from service API, to client, and then to PS/CLI tools. 

I'm not concluding that whether extension 'images' is the right word or concept here, but if the problem exists and needs to be fixed, it shall be fixed at the root place - the service APIs.

@sauryadas Thoughts?

**Comment by ahmetalpbalkan** [comment]
@huangpf "extension images" term is not part of any public documentation and has never been (@squillace can confirm). CLI should be consistent with docs. I also developed extensions and I have not seen "extension images" term anywhere in internal design docs either.

**Comment by huangpf** [comment]
Actually, I already reported to the PIR team that I was also not able to find any public documentation about PIR APIs for ARM VM images & extension images. But the case is that this term has already been published in the other official releases of ASM client, ARM client, and PS. It's probably new to CLI, but it's not a new concept elsewhere.

ASM client:
https://github.com/Azure/azure-sdk-for-net/blob/master/src/ServiceManagement/Compute/ComputeManagement/Generated/IExtensionImageOperations.cs#L35

ARM client:
https://github.com/Azure/azure-sdk-for-net/blob/master/src/ResourceManagement/Compute/ComputeManagement/Generated/VirtualMachineExtensionImageOperations.cs#L40

PS:
https://msdn.microsoft.com/en-us/library/mt150024.aspx

===========================

Again, the change is just to name a command referring to a term by the service API. Renaming or redesigning it shall be another topic.

Would ```extension types``` vs ```extension``` be less confusing? We value your input to choose a right word to distinguish the two different items. Thanks.

**Comment by squillace** [comment]
@huangpf I appreciate the research. We're still missing the point, perhaps because the CLI is used by end-user sysops and devs, and they are looking for "VM Extensions". Why? Because that's what our [docs say](https://azure.microsoft.com/en-us/documentation/articles/virtual-machines-extensions-features/). The fact that our REST api has "images" in the operation name is great for the REST api; end users have NEVER been told to find the right "extension image", ever. Even in the PS version, the doc says:

```
Get-AzureVMExtensionImage
Gets all versions for an Azure extension.
```

In other words, although they USED the same dumb, confusing name, they then went and told the USER to ignore the "image" part because, what does it return? 

It returns VERSIONS for what? An "image"? NO. an EXTENSION.

I'm going to say this again. The documentation tells the user to ignore the name of the cmdlet, because it doesn't do anything like what the name says at all. The documentation says what the customer expects. The customer is looking for the versions of an extension. The cmdlet should be called 'Get-AzureVMExtensionVersions`, but they were too lazy.

We have to help our users. I'm fine if you think this is a UX issue, or a design issue, or whatever, but we need to be clear that the design of the CLI is for users of the cli, NOT for the designers of the ARM REST operation names.

**Comment by squillace** [comment]
Then there's the other issue, which is the list-publishers/offers crap. And by "Crap" I mean "crap", in that I cannot find a single customer this helps, rather than hurts. This is, like the "image" issue, a mirror of the same issue with `azure vm image list-publishers`, which is the way the REST api was built and which is a complete mistake and waste of time and which is going to be fixed. I'd get it if you left that part alone until it is, but I want `image` out of the command name because it's completely wrong. `type` is wrong. `image` is wrong. we make and offer third-party `extensions` and that's it. Remember, it doesn't matter to the caller what "type" -- in the object-orientation sense -- we're using; to the customer, these are all just strings. Everything from their point of view is a string. There are no types. Even "versions" are just different strings.

**Comment by huangpf** [comment]
@squillace I totally understand your concerns and I would agree with you that we don't have to re-use the same name from the REST API. But just FYI, there is another goal to unify the UX of client, PS, & CLI, as much as possible, because these are going to be a more cohesive bundle release soon. (BTW, that PS cmdlet documentation doesn't look accurate, and the PIR documentation for listing extension 'images' is not published publicly yet. So, no official documentation reference yet.)

First of all, we would like to make sure all functionalities are unblocked.

Then, for the correct UX & design, I think the best strategy here would be for @sauryadas to hold a discussion/design meeting and finalize what's the best terms that can shall be used here for CLI, as well as other related client/tool users. FYI, here are some options just for comparison:
```
extension image vs extension
extension artifact vs extension
extension type vs extension
available extension vs extension
platform extension vs extension
extension vs extension
...
```

Last but not least, your suggestion of not distinguishing these may also introduce confusions, i.e. ```vm extension list``` is ambiguous between listing extensions on VMs and the 'image' ones.

**Comment by ahmetalpbalkan** [comment]
@huangpf Can you point me to an example of "extension image" or "extension artifact"? I don't think we don't know what those are.

If you want to disambiguate listing extensions on the VM and list available extensions, here's a simple idea:

* `azure vm-extensions list`: lists available `vm-extensions`
* `azure vm get-extensions <vm>`: shows extensions on the VM.
* `azure vm set-extension <vm>`: sets extension to the VM
* `azure vm remove-extension <vm>`: removes extension from the VM

**Comment by squillace** [comment]
@huangpf Agree on the coming design. I'll be meeting with him and mohit and david this coming week to lay out a few principles on such things so that they can make good decisions. If @ahmetalpbalkan's nice I may invite him. :-)

About the ambiguity, are you discussing the difference between extensions that the current may see or have access to and the extensions that exist but which they may not use given their current subscription?

**Comment by huangpf** [comment]
@ahmetalpbalkan Please refer to the links I mentioned above for ```image```. ```artifact``` is just one of the possible options, referring to the concept of VM image artifact and the REST API (```ArtifactTypes``` in ```"id":"/Subscriptions/xxx/Providers/Microsoft.Compute/Locations/westus/Publishers/Microsoft.Compute/ArtifactTypes/VMExtension/Types/BGInfo/Versions/2.1"```), which is kind of like a class name in the code.

@squillace A meeting among us sounds great. The ambiguity is between: 1) the list of available extension image/type/artifact/... that can be added to a VM, which is functionally like a VM image or gallery item and it's not an ARM resource, and 2) the list of added extensions on a VM, which is an ARM sub-resource under the VM.

Examples -- extension image/type/artifact/... item
```
  {
    "id": "/Subscriptions/xxx/Providers/Microsoft.Compute/Locations/westus/Publishers/Microsoft.Compute/ArtifactTypes/VMExtension/Types/BGInfo/Versions/2.1",
    "name": "2.1",
    "location": "westus",
    "publisher": "Microsoft.Compute",
    "typeName": "BGInfo"
  }
```

vs. an ARM sub-resource extension:
```
    {
      "properties": {
        "publisher": "Microsoft.Compute",
        "type": "BGInfo",
        "typeHandlerVersion": "2.1",
        "autoUpgradeMinorVersion": false,
        "provisioningState": "Succeeded"
      },
      "id": "/subscriptions/xxx/resourceGroups/xxx/providers/Microsoft.Compute/virtualMachines/test/extensions/BGInfo",
      "name": "BGInfo",
      "type": "Microsoft.Compute/virtualMachines/extensions",
      "location": "westus"
    }
```

**Comment by ahmetalpbalkan** [comment]
Alright, I'll withdraw from this conversation. The terminology used in the REST API design is not satisfying either, it is just making things more confusing. I wish we can find a way to avoid REST API to influence the CLI design.

**Comment by squillace** [comment]
@ahmetalpbalkan you bet. 

@huangpf if that's the case, then the solution that Ahmet gave (`azure vm-extensions list`) is actually best. My guess, though, is that you don't want to add another "top level" command. In that case, I'd suggest `azure vm available-extension list`, because:

1. It pushes the command underneath the top level
2. Differentiates between the instances on your vm and the possible instances
3. does not imply that you need to know anything about the REST language, system types or implementations
4. and allows you to leave the `azure vm extension list` command as is

You could switch the order to match the patterns of the other command when you type `azure vm --help` so that it's `azure vm extensions-available list` but that's a bit strange in English. I like Ahmet's choice or the alternative I suggested. I'll follow up on a meeting. Cheers.

**Comment by huangpf** [comment]
Thanks @squillace and @ahmetalpbalkan. I actually kind of like the suggestion of ```vm available-extension``` as for the category name.

I would like to close this thread for now, because since the incoming release, these two different types of extensions will start to belong to different command categories, i.e. ```vm extension``` vs. ```vm extension-image```, as a first step to reduce the confusion. 

Later on, we can decide the suitable naming for them, i.e. possibly ```available-extension```. But let's use another issue to track it. Thanks.

**Comment by huangpf** [comment]
Related PR #2398

```
Lists the virtual machines images
  vm extension-image list [options] <location> [publisher] [type]
```

Listing all images from a location:
```
azure vm extension-image list westus
info:    Executing command vm extension-image list
data:    Publisher                                             Type                                 Version           Location
data:    ----------------------------------------------------  -----------------------------------  ----------------  --------
data:    alertlogic                                            AlertLogicLM                         1.3.0.1           westus
data:    AlertLogic.Extension                                  AlertLogicLM                         1.3.0.0           westus
data:    AlertLogic.Extension                                  AlertLogicLM                         1.4.0.0           westus
data:    AlertLogic.Extension                                  AlertLogicLM                         1.9.0.0           westus
data:    AlertLogic.Extension                                  AlertLogicLM                         1.9.1.0           westus
data:    Chef.Bootstrap.WindowsAzure                           CentosChefClient                     11.12.4.2         westus
data:    Chef.Bootstrap.WindowsAzure                           CentosChefClient                     11.14.6.1         westus
data:    Chef.Bootstrap.WindowsAzure                           ChefClient                           11.10.4           westus
data:    Chef.Bootstrap.WindowsAzure                           ChefClient                           11.12.0.0         westus
data:    Chef.Bootstrap.WindowsAzure                           ChefClient                           11.18.6.2         westus
...
```

Listing all images from a location and publisher:
```
azure vm extension-image list westus alertlogic
```

Listing all images from a location and publisher:
```
azure vm extension-image list westus alertlogic

data:    Publisher   Type          Version  Location
data:    ----------  ------------  -------  --------
data:    alertlogic  AlertLogicLM  1.3.0.1  westus
```

Listing all images from a location, publisher, and type:
```
azure vm extension-image list westus alertlogic AlertLogicLM

data:    Publisher   Type          Version  Location
data:    ----------  ------------  -------  --------
data:    alertlogic  AlertLogicLM  1.3.0.1  westus
```
